### PR TITLE
Make Rate to select the clock to work with

### PR DESCRIPTION
**Pull request by AlexeyMerzlyakov**
This is the continue of [#1373] [discussion](https://github.com/ros2/rclcpp/pull/1373#discussion_r1118830044) and related to the https://github.com/ros-planning/navigation2/issues/3325 issue in Nav2

The initial idea is to make `rclcpp::Rate` to work with ROS-time (currently it works only with system and steady times).

What was made:
* Add `rclcpp::Clock` pointer to the `rclcpp::Rate` constructor to allow developer to specify the clock to work with (by-default it is still `RCL_SYSTEM_TIME` if no clock is specified)
* Add `rclcpp::ROSRate` rate class, respective to ROS-time

Current implementation points:
* Getting rid of templates' structure allows developer to select the clock to work with in dynamics
* Renamed `rclcpp::GenericRate` -> to `rclcpp::Rate` and gave it the default clock to work with
* `rclcpp::Rate` and `rclcpp::WallRate` API remaining to be untouched as much as possible
* Although there are some changes in API: `rclcpp::Duration` instead of `std::chrono::nanoseconds` in durations and `Rate::get_type()` routine instead of previously used `Rate::is_steady()`
* New `rclcpp::ROSRate` class that working with ROS time

### CODE REVIEW

**Comment by AlexeyMerzlyakov** [comment]
The PR contains only the `time.hpp` change. If the change will be agreed, I will fix/add new tests on it.

File: rclcpp/include/rclcpp/rate.hpp
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/rate.hpp:50]: I think we need to deprecate this for this moment with following message.


``[[deprecated("use rclcpp::Rate class instead of GenericRate")]]``

**Review by fujitatomoya — COMMENTED** [review]
overall lgtm

File: rclcpp/include/rclcpp/rate.hpp
Comment (AlexeyMerzlyakov) [reply rclcpp/include/rclcpp/rate.hpp:50]: Done in https://github.com/ros2/rclcpp/pull/2123/commits/c25fe6bdad6ee686fa411a89cd7ae140e0f3ac40 and https://github.com/ros2/rclcpp/pull/2123/commits/7323e9730a8f3e288177bb30b74777f34e33dfee. `GenericRate`'s API remained to be the same in order to make the compiler to inform developer only about class name change. Implementation was cleaned-out.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: rclcpp/include/rclcpp/rate.hpp
Comment (AlexeyMerzlyakov) [reply rclcpp/include/rclcpp/rate.hpp:71]: In vanilla branch, `time_to_sleep` is used to [avoid](https://github.com/ros2/rclcpp/blob/20e9cd17f6723337dc0915f4e2a966f9585c309c/rclcpp/include/rclcpp/rate.hpp#L75) negative values, and was initialized [just before](https://github.com/ros2/rclcpp/blob/20e9cd17f6723337dc0915f4e2a966f9585c309c/rclcpp/include/rclcpp/rate.hpp#L71) the usage. In this patch, `time_to_sleep` became to be of `rclcpp::Duration` type, which can not be negative, so this check was [transformed](https://github.com/AlexeyMerzlyakov/rclcpp/blob/7323e9730a8f3e288177bb30b74777f34e33dfee/rclcpp/include/rclcpp/rate.hpp#L89). `time_to_sleep` became to be required later, so it [was moved](https://github.com/AlexeyMerzlyakov/rclcpp/blob/7323e9730a8f3e288177bb30b74777f34e33dfee/rclcpp/include/rclcpp/rate.hpp#L100) to the place right before usage, for better code readability.

File: rclcpp/include/rclcpp/rate.hpp
Comment (AlexeyMerzlyakov) [reply rclcpp/include/rclcpp/rate.hpp:47]: The purpose to leave `GenericRate` class - is to warn developer that `Rate` should be used instead, from on now. All implementations were left empty to have only one one compiler deprecation warning message and thus to not cause a confusion of other inducted compiler errors (e.g. like "can not find such constructor or method"). It is supposed that developer will switch to `Rate` class after this message, so `GenericRate`'s implementations were emptied.
However, indeed, if warnings were disabled as errors on developer side, and he will refuse to follow the recommendations, this indeed might cause an undefined behavior.

> I'm not sure why this doesn't use `Rate` or `Rate` isn't implemented as `GenericRate` and we typedef a `Rate` from it or something if it doesn't make sense to have dual implementations

In vanilla version, `Rate` is a templated implementation of `GenericRate` (`using Rate = GenericRate<std::chrono::system_clock>;` or as you've mentioned, `typedef GenericRate<std::chrono::system_clock> Rate;` ). However, in current implementation, `Rate` concept now was changed. It is non-templated class, which can not be a template typedef of `GenericRate`. Moreover, if forcibly leave the implementation to use the `Rate` as `typedef` of `GenericRate` (as it now in vanilla), `GenericRate` as deprecated class, will cause compiler to warn when we will try to make the `Rate` as `typedef GenericRate<something>` .

Or did you mean to leave `GenericRate` untouched and have `Rate` as a completely new class, having dual-implementations? Wouldn't it cause even more confusion?

Finally, `GenericRate`'s implementation could be completely removed, which will cause additional compiler errors, but it never allow the developer to build their code with older API and thus to have an undefined behavior.

File: rclcpp/include/rclcpp/rate.hpp
Comment (AlexeyMerzlyakov) [reply rclcpp/include/rclcpp/rate.hpp:102]: The `rclcpp::Clock::sleep_for()` - is a [wrapper](https://github.com/ros2/rclcpp/blob/20e9cd17f6723337dc0915f4e2a966f9585c309c/rclcpp/src/rclcpp/clock.cpp#L179-L183) for `rclcpp::Clock::sleep_until()`. `sleep_until()` seems to have a time [clock handler](https://github.com/ros2/rclcpp/blob/20e9cd17f6723337dc0915f4e2a966f9585c309c/rclcpp/src/rclcpp/clock.cpp#L140), catching time jumps and clock source changes. It seems, that during this change waiting is being interrupted [[1](https://github.com/ros2/rclcpp/blob/20e9cd17f6723337dc0915f4e2a966f9585c309c/rclcpp/src/rclcpp/clock.cpp#L146)], [[2](https://github.com/ros2/rclcpp/blob/20e9cd17f6723337dc0915f4e2a966f9585c309c/rclcpp/src/rclcpp/clock.cpp#L167)].

So, we have two types of situations:
1. Time was jumped (backward or forward) during the waiting:
  - In this case this situation will be handled correctly by the `rclcpp::Clock::sleep_until()`.
  One thing it looks I've missed: `sleep()` routine needs return the value from `sleep_for()` instead of always `true`:
  ```
  - clock_->sleep_for(time_to_sleep);
  - return true;
  + return clock_->sleep_for(time_to_sleep);
  ```
2. Time was jumped between `sleep()` beginning (when we already got `now` variable) and `clock_->sleep_for()` call. Very rare situation, but anyway:
 - In this case, inside `sleep_for()` the time to sleep will be [calculated](https://github.com/ros2/rclcpp/blob/20e9cd17f6723337dc0915f4e2a966f9585c309c/rclcpp/src/rclcpp/clock.cpp#L182) in already new time after jump, and we will sleep a required `time_to_sleep`, but in new time. Not fully correct situaion, but later on, on next `sleep()` call, if time was jumped (back or forward) more than on one `period_`, it will be handled inside the `sleep()`: for [back](https://github.com/AlexeyMerzlyakov/rclcpp/blob/7323e9730a8f3e288177bb30b74777f34e33dfee/rclcpp/include/rclcpp/rate.hpp#L82) jump and for [forward](https://github.com/AlexeyMerzlyakov/rclcpp/blob/7323e9730a8f3e288177bb30b74777f34e33dfee/rclcpp/include/rclcpp/rate.hpp#L89) jump; and situation will be stabilized.

File: rclcpp/include/rclcpp/rate.hpp
Comment (SteveMacenski) [reply rclcpp/include/rclcpp/rate.hpp:47]: I don't think that's the right philosophy to have on deprecation warnings. The warning is to let people know it will be depreciated soon in future releases, not that what you're currently using is already deprecated and will not work.

Comment (clalancette) [reply rclcpp/include/rclcpp/rate.hpp:47]: > I don't think that's the right philosophy to have on deprecation warnings. The warning is to let people know it will be depreciated soon in future releases, not that what you're currently using is already deprecated and will not work.

Yes, agreed.  The old API needs to continue to work (with a compile-time deprecation warning).

File: rclcpp/include/rclcpp/rate.hpp
Comment (AlexeyMerzlyakov) [reply rclcpp/include/rclcpp/rate.hpp]: > I'll do it in next PR update.

Done in the [latest commit](https://github.com/ros2/rclcpp/pull/2123/commits/30b95de7cc242e1981091f406fb9c8226d34f97f#diff-c6ecfbeee7b7384c9e72a8984cc2fc718415b3bb741c0ea41ffc1d9b0718fea7R164)

File: rclcpp/include/rclcpp/rate.hpp
Comment (AlexeyMerzlyakov) [reply rclcpp/include/rclcpp/rate.hpp]: Okay, got it. I've restored the previous functionality of `GenericRate` class and API (including `is_steady()`) in the latest update

**Comment by AlexeyMerzlyakov** [comment]
The build was failed due to outdated branch. Rebased over latest `rolling` and force-pushed branch.

File: rclcpp/include/rclcpp/rate.hpp
Comment (SteveMacenski) [inline rclcpp/include/rclcpp/rate.hpp:48]: Shouldn't this derive from RateBase again?

**Review by SteveMacenski — APPROVED** [review]
Otherwise LGTM

File: rclcpp/include/rclcpp/rate.hpp
Comment (AlexeyMerzlyakov) [reply rclcpp/include/rclcpp/rate.hpp:48]: Since of meaning of clock types has been updated (ROS-respecitve clock type added), new `rcl_clock_type_t get_type()` method [was introduced](https://github.com/ros2/rclcpp/pull/2123/files#diff-c6ecfbeee7b7384c9e72a8984cc2fc718415b3bb741c0ea41ffc1d9b0718fea7R39) in a `RateBase` basic class as a alternative/replacement for `bool is_steady()` one (I would like to completely remove `is_steady()` method from on now as unnecessary, but since we need to support old API for compativility for `rclcpp::Rate` and `rclcpp::WallRate` classes, `is_steady` was remain). `GenericRate` if being a child-class of `RateBase` now should implement `get_type()` method. But since `GenericRate` - is deprecated class, it might be overkill, I believe.

Therefore `GenericRate` was taken out of parenthood of `RateBase` remaining the same functionality as it previously has; while new `RateBase -> Rate -> WallRate/ROSRate` lineage was introduced instead.

**Comment by clalancette** [comment]
Here is CI (just on Linux) to see how much we would have to change to get this in:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18580/)

**Comment by clalancette** [comment]
@ros-pull-request-builder retest this please

**Comment by clalancette** [comment]
So there were a *lot* of failures in CI.  That may be because this needs to be rebased, I'm not sure.  @AlexeyMerzlyakov @SteveMacenski can you rebase this onto the latest and we can try again?

**Comment by wjwwood** [comment]
@Mergifyio rebase

**Comment by wjwwood** [comment]
@Mergifyio rebase

**Comment by wjwwood** [comment]
New CI after rebase: [Build Status](https://ci.ros2.org/job/ci_linux/18587/)

**Comment by SteveMacenski** [comment]
@AlexeyMerzlyakov take a look. The policy is that deprecations in the core need to have PRs to resolve so that builds are green without deprecated API use. So anywhere that we have failures we should have complimentary PRs in the ROS 2 core so that everything turns over OK. Unfortunately due to that delay, this won't be able to get in before Iron's freeze

**Comment by SteveMacenski** [comment]
Any update?

**Comment by AlexeyMerzlyakov** [comment]
Waiting until https://github.com/ros2/system_tests/pull/516 (that covers failed tests) to be reviewed on the way to its merge. Then we need to restart testing again in this PR to make sure that all regressions were resolved.

File: rclcpp/include/rclcpp/rate.hpp
Comment (fujitatomoya) [reply rclcpp/include/rclcpp/rate.hpp:48]: Sorry to be late to get back to you. What we need to deprecate is `RateBase::is_steady` since the application can get the clock type?

**Review by fujitatomoya — COMMENTED** [review]
overall lgtm! just one minor comment.

File: rclcpp/include/rclcpp/rate.hpp
Comment (clalancette) [reply rclcpp/include/rclcpp/rate.hpp:48]: The problem with removing `GenericRate` from the hierarchy is that this breaks certain applications.  Consider for example if a downstream node has `std::vector<RateBase>` in their code.  In that case, they will no longer be able to store a `GenericRate` in there, since it is no longer a child class of `RateBase`.

So I think we should add this back to the hierarchy (i.e. make this a child of `RateBase`), and implement `get_type()` as returning either `RCL_SYSTEM_TIME` (if `Clock::is_steady` is false), or `RCL_STEADY_TIME` (if `Clock::is_steady` is true).

Comment (AlexeyMerzlyakov) [reply rclcpp/include/rclcpp/rate.hpp]: @fujitatomoya, OK. Thanks for reaching me. I've made `is_steady()` method to be deprecated for all classes in [this](https://github.com/ros2/rclcpp/pull/2123/commits/5f08e09bb048558689fb4e54e28b94bd1f9340ca) commit.

@clalancette, Got your point. This is indeed makes sense, so `RateBase` -> `GenericRate` relationship was restored in the [same](https://github.com/ros2/rclcpp/pull/2123/commits/5f08e09bb048558689fb4e54e28b94bd1f9340ca) commit.

**Comment by AlexeyMerzlyakov** [comment]
For the clarity, I've prepared API `Rate*` change briefly summarized in the following document:
[Rate_API_change.pdf](https://github.com/ros2/rclcpp/files/12171335/Rate_API_change.pdf)

Hope, this might help to take an overall look on this change.

File: rclcpp/include/rclcpp/rate.hpp
Comment (clalancette) [inline rclcpp/include/rclcpp/rate.hpp:169]: Given that this class is not a template, I'd like to see all of this implementation moved into a `rate.cpp` file.  That will help with compile times of downstream projects.

The same goes for all of the other methods below.

**Review by clalancette — CHANGES_REQUESTED** [review]
Thank you so much for putting together the document explaining the change.  It is really helpful.

I've left one change inline.

Besides that, the fact that the `Rate` and `WallRate` classes now require a context to be initialized is a big change from the previous implementation.  That doesn't necessarily mean we shouldn't do this, but I do want to have a conversation about it with the maintainers.  @fujitatomoya @alsora @wjwwood what do you think about this change in semantics?

**Comment by fujitatomoya** [comment]
> Thank you so much for putting together the document explaining the change. It is really helpful.
> 
> I've left one change inline.
> 
> Besides that, the fact that the `Rate` and `WallRate` classes now require a context to be initialized is a big change from the previous implementation. That doesn't necessarily mean we shouldn't do this, but I do want to have a conversation about it with the maintainers. @fujitatomoya @alsora @wjwwood what do you think about this change in semantics?

that is true, after this PR, context must be initialized before using those classes. i think that would not be a problem for ROS 2 applications. and depending on ROS 2 `Clock` objects makes sense to me to support `ROSRate`.

**Comment by alsora** [comment]
The PR looks good to me.
Before merging we should address the comment here (https://github.com/ros2/rclcpp/pull/2123#discussion_r1274959585) about moving stuff to the cpp file.

**Comment by fujitatomoya** [comment]
@AlexeyMerzlyakov friendly ping. i will restart the CI all comments are resolved.

File: rclcpp/include/rclcpp/rate.hpp
Comment (AlexeyMerzlyakov) [reply rclcpp/include/rclcpp/rate.hpp]: Moved all `Rate` implementations into source-file in https://github.com/ros2/rclcpp/pull/2123/commits/3d7b2232dca56ddd1021c174a4617714bb2f9538

**Comment by AlexeyMerzlyakov** [comment]
Apologies for late respond. Yes, this definitely makes sense: moved all non-template implementations to separate source-file in the following commit: https://github.com/ros2/rclcpp/pull/2123/commits/3d7b2232dca56ddd1021c174a4617714bb2f9538, to save compile time of all dependent packages/projects.

@fujitatomoya, before CI-testing restart we need to ensure that https://github.com/ros2/system_tests/pull/516 is in local test container (if it possible to do this).

**Review by clalancette — COMMENTED** [review]
This looks good to me!  Thank you for all of the iterations.  I'm going to run CI on it (along with https://github.com/ros2/system_tests/pull/516) on a number of platforms and see where we are.

**Review by fujitatomoya — APPROVED** [review]
with a couple of nitpicks, lgtm.

**Comment by clalancette** [comment]
Also, this likely needs a rebase onto the latest before we can run CI on it.

File: rclcpp/include/rclcpp/rate.hpp
Comment (AlexeyMerzlyakov) [reply rclcpp/include/rclcpp/rate.hpp]: It was initially copied from `GenericRate` logic, but yes, this definitely makes sense. [Added](https://github.com/ros2/rclcpp/pull/2123/commits/8f19f28dfb72c50ab1d353026ffdf366777f327f) constructors argument checks at least for new `Rate` class. Do we need to add similar checks for obsolete `GenericRate` constructors, or better leave it as-is, making changes only in working API?

File: rclcpp/src/rclcpp/rate.cpp
Comment (AlexeyMerzlyakov) [reply rclcpp/src/rclcpp/rate.cpp]: Done in https://github.com/ros2/rclcpp/pull/2123/commits/8f19f28dfb72c50ab1d353026ffdf366777f327f

**Comment by AlexeyMerzlyakov** [comment]
Force-pushed changes over the latest `rolling` commit. Also force-pushed the changes in https://github.com/ros2/system_tests/pull/516 over latest commit. So it should be ready for round 2 of CI testing.

**Review by fujitatomoya — APPROVED** [review]

File: rclcpp/include/rclcpp/rate.hpp
Comment (clalancette) [inline rclcpp/include/rclcpp/rate.hpp:146]: I think we should move these constructors into rate.cpp as well.

File: rclcpp/include/rclcpp/rate.hpp
Comment (clalancette) [inline rclcpp/include/rclcpp/rate.hpp:193]: And let's move these constructors to rate.cpp as well (sorry I'm being pedantic here, but I'm currently on a quest to reduce compile times in rclcpp and this is one way to help with that).

**Review by clalancette — CHANGES_REQUESTED** [review]
I've left a couple more small things to fix, then I think this will be good.

As far as:

```
Do we need to add similar checks for obsolete GenericRate constructors, or better leave it as-is, making changes only in working API?
```

I'm fine with only adding the checks to the new API, and leaving `GenericRate` as-is.

File: rclcpp/include/rclcpp/rate.hpp
Comment (AlexeyMerzlyakov) [reply rclcpp/include/rclcpp/rate.hpp]: Done in https://github.com/ros2/rclcpp/pull/2123/commits/a947972f50007dd3e0249f71f92017fcf22ca9e0

File: rclcpp/include/rclcpp/rate.hpp
Comment (AlexeyMerzlyakov) [reply rclcpp/include/rclcpp/rate.hpp]: Yep, no problems. Done in last commit

**Comment by clalancette** [comment]
Here's another round of CI to see where we are with this:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19401/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13930/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20122/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/20123/)
* RHEL [Build Status](http://ci.ros2.org/job/ci_linux-rhel/463/)
* Clang [Build Status](http://ci.ros2.org/job/ci_linux_clang_libcxx/29/)

File: rclcpp/include/rclcpp/rate.hpp
Comment (clalancette) [inline rclcpp/include/rclcpp/rate.hpp:123]: Because this is no longer templated, on Windows I think we need to add the `PUBLIC` keyword:

```suggestion
class RCLCPP_PUBLIC Rate : public RateBase
```

Comment (clalancette) [reply rclcpp/include/rclcpp/rate.hpp:123]: (if that doesn't work on Windows, then the other option is to add `RCLCPP_PUBLIC` to each of the methods in the class)

File: rclcpp/include/rclcpp/rate.hpp
Comment (clalancette) [inline rclcpp/include/rclcpp/rate.hpp:160]:
```suggestion
class RCLCPP_PUBLIC WallRate : public Rate
```

File: rclcpp/include/rclcpp/rate.hpp
Comment (clalancette) [inline rclcpp/include/rclcpp/rate.hpp:167]:
```suggestion
class RCLCPP_PUBLIC ROSRate : public Rate
```

**Review by clalancette — CHANGES_REQUESTED** [review]

**Review by SteveMacenski — APPROVED** [review]
Beyond comments that were made, this looks good to me!

File: rclcpp/test/rclcpp/test_rate.cpp
Comment (AlexeyMerzlyakov) [reply rclcpp/test/rclcpp/test_rate.cpp:406]: Yes, there is needed macro in `utils/rclcpp_gtest_macros.hpp`. Switched to it in the following commit. Thanks for notice!

**Comment by AlexeyMerzlyakov** [comment]
Oops, seems like GitHub auto-commits broken the DCO signature test. Rebasing it with adding the signature and force-pushing the branch.

**Comment by clalancette** [comment]
Let's see if Windows is happier with this:

* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20143/)

**Comment by clalancette** [comment]
The good news is that it looks like we got beyond the compiling phase on Windows.  So I'm optimistic that this will be OK there.  I'm going to kick of the rest of CI here to parallelize things:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19422/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13950/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20143/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/20145/)
* RHEL [Build Status](http://ci.ros2.org/job/ci_linux-rhel/465/)
* Clang [Build Status](http://ci.ros2.org/job/ci_linux_clang_libcxx/30/)

**Comment by clalancette** [comment]
So we are getting closer, but unfortunately Windows says:

```
non dll-interface class 'rclcpp::RateBase' used as base for dll-interface class 'rclcpp::Rate' (compiling source file C:\ci\ws\src\ros2\demos\action_tutorials\action_tutorials_cpp\src\fibonacci_action_server.cpp)
```

I'm honestly not sure what that means; that is a new one for me.  Maybe we need `RCLCPP_PUBLIC` on `RateBase`?  That's not a pattern I'm familiar with, but that may be what it means.

**Comment by clalancette** [comment]
OK, I *think* my latest commit (bff4fd0) should fix things on Windows.  Let's run CI again and see if that is the case:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19424/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13952/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20147/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/20148/)
* RHEL [Build Status](http://ci.ros2.org/job/ci_linux-rhel/466/)
* Clang [Build Status](http://ci.ros2.org/job/ci_linux_clang_libcxx/31/)

**Comment by AlexeyMerzlyakov** [comment]
@clalancette, thank you for combing-up Windows builds!
The latest Windows build fail seems really strange and related to some CI / cache issue:
```
03:59:06   * seven_zip_archive[iconv] action extractc:/temp/local-mode-cache/cache/cookbooks/windows/libraries/windows_helper.rb:74:in `cached_file': URI.unescape is obsolete (StructuredWarnings::StandardWarning)
```

However, Clang builds give linking problems, seem to be related to the latest commit
```
04:37:36 [  8%] Building CXX object CMakeFiles/rclcpp.dir/src/rclcpp/any_executable.cpp.o
04:37:36 In file included from /home/jenkins-agent/workspace/ci_linux_clang_libcxx/ws/src/ros2/rclcpp/rclcpp/src/rclcpp/any_executable.cpp:15:
04:37:36 In file included from /home/jenkins-agent/workspace/ci_linux_clang_libcxx/ws/src/ros2/rclcpp/rclcpp/include/rclcpp/any_executable.hpp:20:
04:37:36 In file included from /home/jenkins-agent/workspace/ci_linux_clang_libcxx/ws/src/ros2/rclcpp/rclcpp/include/rclcpp/callback_group.hpp:30:
04:37:36 In file included from /home/jenkins-agent/workspace/ci_linux_clang_libcxx/ws/src/ros2/rclcpp/rclcpp/include/rclcpp/timer.hpp:31:
04:37:36 /home/jenkins-agent/workspace/ci_linux_clang_libcxx/ws/src/ros2/rclcpp/rclcpp/include/rclcpp/rate.hpp:43:3: error: an attribute list cannot appear here
04:37:36   [[deprecated("use get_type() instead")]] virtual bool is_steady() const = 0;
```

Unfortunately, I checked only Linux builds when prepared the commit, so missed this point. If everything will fail, on next week I'll try to get out with Windows builds and check the problem as well

**Comment by clalancette** [comment]
> The latest Windows build fail seems really strange and related to some CI / cache issue:

Yeah, it looks like the network was having problems there.  I'll kick them off again.

> However, Clang builds give linking problems (is it related to the latest commit?)

Yeah, most probably.  We'll have to take a closer look at building with clang.

**Comment by clalancette** [comment]
I think f54ca7d should fix it on clang.  Let's see if that is happy across the board:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19425/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13953/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20151/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/20152/)
* RHEL [Build Status](http://ci.ros2.org/job/ci_linux-rhel/467/)
* Clang [Build Status](http://ci.ros2.org/job/ci_linux_clang_libcxx/33/)

**Review by clalancette — APPROVED** [review]
Thanks for all of the back-and-forth here @AlexeyMerzlyakov .  With the latest changes, this looks good to me.  I'm going to go ahead and merge, along with https://github.com/ros2/system_tests/pull/516

**Comment by clalancette** [comment]
Just as an FYI, I released this into Rolling yesterday and I think there was some downstream fallout from this change.  In particular, look at:

* https://build.ros2.org/view/Rbin_uJ64/job/Rbin_uJ64__mavros__ubuntu_jammy_amd64__binary/233/console
* https://build.ros2.org/view/Rbin_uJ64/job/Rbin_uJ64__topic_tools__ubuntu_jammy_amd64__binary/91/console
* https://build.ros2.org/view/Rbin_uJ64/job/Rbin_uJ64__moveit_servo__ubuntu_jammy_amd64__binary/231/console

All of those failures revolve around `rclcpp::Rate`, which makes me suspect this.  I'm going to take a look, but @AlexeyMerzlyakov it would be helpful if you could look as well.

**Comment by clalancette** [comment]
OK, I found it.  See #2301 which fixes it.

**Pull request merged**
